test/core/end2end/tests/connectivity.cc

```diff
@@ -163,9 +169,74 @@ static void test_connectivity(grpc_end2end_test_config config) {
   cq_verifier_destroy(cqv);
 }
 
+static void cb_watch_connectivity(grpc_experimental_completion_queue_functor *functor, int success) {
+  callback_context* cb_ctx = (callback_context*) functor;
+
+  gpr_log(GPR_DEBUG, "cb_watch_connectivity called, verifying");
+
+  /* callback must not have errors */
+  GPR_ASSERT(success != 0);
+
+  gpr_mu_lock(&cb_ctx->mu);
+  cb_ctx->done = true;
+  gpr_mu_unlock(&cb_ctx->mu);
+}
+
+static void cb_shutdown(grpc_experimental_completion_queue_functor *functor, int success) {
+  gpr_log(GPR_DEBUG, "cb_shutdown called, nothing to do");
+}
+
+static void test_watch_connectivity_cq_callback(grpc_end2end_test_config config) {
+  grpc_experimental_completion_queue_functor shutdown_functor;
+  callback_context cb_ctx;
+  grpc_completion_queue * cq;
+  grpc_end2end_test_fixture f = config.create_fixture(nullptr, nullptr);
+
+  grpc_channel_args client_args;
+  client_args.args = nullptr;
+  client_args.num_args = 0;
+
+  config.init_client(&f, &client_args);
+
+  shutdown_functor.functor_run = cb_shutdown;
+
+  /* initalize the callback context */
+  cb_ctx.functor.functor_run = cb_watch_connectivity;
+  cb_ctx.done = false;
+  gpr_mu_init(&cb_ctx.mu);
+
+  /* start connecting */
+  grpc_channel_check_connectivity_state(f.client, 1);
+
+  /* create the cq callback */
+  cq = grpc_completion_queue_create_for_callback(&shutdown_functor, NULL);
+
+  /* start watching for any change, cb is immediately called
+   * and no dead lock should be raised */
+  grpc_channel_watch_connectivity_state(f.client, GRPC_CHANNEL_IDLE,
+                                        grpc_timeout_seconds_to_deadline(3),
+                                        cq, &cb_ctx.functor);
+
+  /* we just check that the callback was executed once notifying the transition
```

Comment:
Something seems off here.  Do we have a guarantee in this test that the callback will have been invoked before `grpc_channel_watch_connectivity_state()` returns?  If not, then we probably need to use a `gpr_event` to wait for the callback to run here; if so, then we probably don't need `cb_ctx.done` to be protected by synchronization in the first place.

Reply:
Theoretically taking into account that we are using as a last connection status the `GRPC_CHANNEL_IDLE` status and trying a connection in two lines before the callback should be called immediately, otherwise the test won't pass.

Initially, I thought that I would need some kind of synchronization - here the reason of having the mutex - but then I realized that it was not really necessary because the callback was immediately called within the execution of the `grpc_channel_watch_connectivity_state` function.

My fault that I didn't remember to get rid of the mutex variable, which is not really needed.

True that another way to implement this is by using a `gpr_event`, but in some way you will be forced of checking if the `gpr_event_wait` finishes because of the callback signal or because of a timeout, that most likely will force you of having to use a kind of `done` variable. Well, you always have the chance of not using a timeout I guess :)

I'm still considering that the current approx is enough and IMO simpler taking into account how the callback is being called within the context execution of the `grpc_channel_watch_connectivity_state ` function.

Thoughts?